ndindex/array.py

```diff
@@ -0,0 +1,122 @@
+import warnings
+
+from numpy import ndarray, asarray, integer, bool_, array2string, empty, intp
+
+from .ndindex import NDIndex, asshape
+
+class ArrayIndex(NDIndex):
+    """
+    Superclass for array indices
+
+    This class should not be instantiated directly. Rather, use one of its
+    subclasses, :class:`~.IntegerArray` or :class:`~.BooleanArray`.
+
+    To subclass this, define the `dtype` attribute, as well as all the usual
+    ndindex methods.
+    """
+    # Subclasses should redefine this
+    dtype = None
+
+    def _typecheck(self, idx, shape=None):
+        if self.dtype is None:
+            raise TypeError("Do not instantiate the superclass ArrayIndex directly")
+
+        if shape is not None:
+            if idx != []:
+                raise ValueError("The shape argument is only allowed for empty arrays (idx=[])")
+            shape = asshape(shape)
+            if 0 not in shape:
+                raise ValueError("The shape argument must be an empty shape")
+            idx = empty(shape, dtype=self.dtype)
+
+        if isinstance(idx, (list, ndarray, bool, integer, int, bool_)):
+            # Ignore deprecation warnings for things like [1, []]. These will be
+            # filtered out anyway since they produce object arrays.
+            with warnings.catch_warnings(record=True):
+                a = asarray(idx)
+                if a is idx:
+                    a = a.copy()
+                if isinstance(idx, list) and 0 in a.shape:
+                    a = a.astype(self.dtype)
+            if self.dtype == intp and issubclass(a.dtype.type, integer):
+                if a.dtype != self.dtype:
+                    a = a.astype(self.dtype)
+            if a.dtype != self.dtype:
+                raise TypeError(f"The input array to {self.__class__.__name__} must have dtype {self.dtype.__name__}, not {a.dtype}")
+            a.flags.writeable = False
+            return (a,)
+        raise TypeError(f"{self.__class__.__name__} must be created with an array with dtype {self.dtype.__name__}")
+
+    @property
+    def raw(self):
```

Comment:
Should raw be an alias to array?

Reply:
The advantage of this way is it will get the docstring of `NDIndex.raw`, which explains the `raw` attribute generally for all ndindex objects.

Reply:
Ahh fair